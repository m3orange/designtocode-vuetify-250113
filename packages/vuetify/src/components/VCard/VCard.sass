// Imports
@import './index'

.v-card
  background: $card-background
  color: $card-color
  position: relative

  &::before,
  &::after
    border-radius: inherit
    bottom: 0
    content: ''
    display: block
    left: 0
    pointer-events: none
    position: absolute
    right: 0
    top: 0
    transition-property: box-shadow, opacity
    transition: 0.28s $standard-easing

  &::before
    @include elevation($card-elevation)

  @include border($card-border...)
  @include position($card-positions)
  @include rounded($card-border-radius)

  &--border
    border-width: $card-border-thin-width

  &--disabled
    pointer-events: none
    user-select: none

    > [class*='v-card-']
      opacity: $card-disabled-opacity

  &--hover
    cursor: pointer

    &::before
      opacity: 1
      z-index: -1

    &::after
      z-index: 1
      opacity: 0

      @include elevation($card-hover-elevation)

    &:hover::after
      opacity: 1

    &:hover::before
      opacity: 0

  &--tile
    @include rounded($card-tile-border-radius)

.v-card-title
  align-items: center
  display: flex
  font-size: $card-title-font-size
  font-weight: $card-title-font-weight
  letter-spacing: $card-title-letter-spacing
  text-transform: $card-title-text-transform

  @include card-line-height-densities($card-title-densities)

.v-card-subtitle
  align-items: center
  color: $card-subtitle-color
  display: flex
  font-size: $card-subtitle-font-size
  font-weight: $card-subtitle-font-weight
  letter-spacing: $card-subtitle-letter-spacing
  text-transform: $card-subtitle-text-transform

  @include card-line-height-densities($card-subtitle-densities)

.v-card-text
  color: $card-text-color
  font-size: $card-text-font-size
  font-weight: $card-text-font-weight
  letter-spacing: $card-text-letter-spacing
  padding: $card-text-padding
  text-transform: $card-text-text-transform

  @include card-line-height-densities($card-text-densities)

.v-card-actions
  align-items: center
  display: flex
  flex: 1 1 auto
  min-height: $card-actions-min-height
  padding: $card-actions-padding

.v-card-media
  overflow: hidden

  &:first-child
    .v-card &
      border-top-left-radius: inherit
      border-top-right-radius: inherit

  &:last-child
    .v-card &
      border-bottom-left-radius: inherit
      border-bottom-right-radius: inherit

.v-card-img
  border-radius: inherit
  display: flex
  height: 100%
  left: 0
  overflow: hidden
  position: absolute
  top: 0
  width: 100%
  z-index: -1

.v-card-avatar
  @include ltr()
    margin-right: $card-avatar-margin-start

  @include rtl()
    margin-left: $card-avatar-margin-end

.v-card-header
  flex: 1 1 auto

.v-card-item
  align-items: center
  display: flex
  padding: $card-item-padding
  height: $card-item-height

  .v-card--density-comfortable &
    height: $card-item-comfortable-height

  .v-card--density-compact &
    height: $card-item-compact-height
