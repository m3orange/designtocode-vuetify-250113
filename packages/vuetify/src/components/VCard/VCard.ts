// Styles
import '../../stylus/components/_cards.styl'

// Extensions
import VSheet from '../VSheet'

// Extensions
import VHover from '../VHover'

// Mixins
import Routable from '../../mixins/routable'

// Helpers
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'

/* @vue/component */
export default mixins(
  Routable,
  VSheet
).extend({
  name: 'v-card',

  props: {
    elevation: {
      type: [Number, String],
      default: 2
    },
    flat: Boolean,
    hover: [Boolean, String, Number],
    img: String,
    raised: Boolean
  },

  data: () => ({
    isHovered: false
  }),

  computed: {
    classes (): object {
      return {
        'v-card': true,
        'v-card--hover': this.isHoverable,
        ...VSheet.options.computed.classes.call(this)
      }
    },
    computedElevation (): number | string {
      if (this.isHovered) {
        return typeof this.hover === 'boolean' ? 8 : Number(this.hover)
      }
      if (this.flat) return 0
      if (this.raised) return 3

      return (VSheet.options.computed as any).computedElevation.call(this)
    },
    isHoverable (): boolean {
      return this.hover != null && this.hover !== false
    },
    styles (): object {
      const style = {
        ...VSheet.options.computed.styles.call(this)
      }

      if (this.img) {
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h): VNode {
    const { tag, data } = this.generateRouteLink(this.classes)

    data.style = this.styles

    const card = h(tag, this.setBackgroundColor(this.color, data), this.$slots.default)

    if (!this.isHoverable) return card

    return h(VHover, {
      props: { value: this.isHovered },
      on: {
        input: (val: boolean) => {
          this.isHovered = val
        }
      }
    }, [card])
  }
})
