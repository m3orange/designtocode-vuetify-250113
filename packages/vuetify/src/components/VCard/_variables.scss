// Defaults
$card-actions-min-height: 52px !default;
$card-actions-padding: 8px !default;
$card-avatar-margin-end: 1rem !default;
$card-avatar-margin-start: 1rem !default;
$card-background: var(--v-theme-surface) !default;
$card-border-color: $border-color-root !default;
$card-border-elevation: 0 !default;
$card-border-radius: $border-radius-root !default;
$card-border-style: $border-style-root !default;
$card-border-thin-width: thin !default;
$card-border-width: 0 !default;
$card-color: var(--v-theme-on-surface) !default;
$card-disabled-opacity: 0.6 !default;
$card-elevation: 1 !default;
$card-hover-elevation: 8 !default;
$card-item-comfortable-height: 64px !default;
$card-item-compact-height: 56px !default;
$card-item-height: 72px !default;
$card-item-padding: .5rem 1rem !default;
$card-positions: absolute fixed !default;
$card-subtitle-color: rgba(var(--v-theme-on-surface), var(--v-medium-emphasis-opacity)) !default;
$card-subtitle-comfortable-line-height: 1.125rem !default;
$card-subtitle-compact-line-height: 1rem !default;
$card-subtitle-font-size: map-deep-get($typography, 'body-2', 'size') !default;
$card-subtitle-font-weight: map-deep-get($typography, 'body-2', 'weight') !default;
$card-subtitle-letter-spacing: map-deep-get($typography, 'body-2', 'letter-spacing') !default;
$card-subtitle-line-height: map-deep-get($typography, 'body-2', 'line-height') !default;
$card-subtitle-padding: 0 16px !default;
$card-subtitle-text-transform: none !default;
$card-text-color: rgba(var(--v-theme-on-background), var(--v-medium-emphasis-opacity)) !default;
$card-text-comfortable-line-height: 1.2rem !default;
$card-text-compact-line-height: 1.15rem !default;
$card-text-font-size: map-deep-get($typography, 'body-2', 'size') !default;
$card-text-font-weight: map-deep-get($typography, 'body-2', 'weight') !default;
$card-text-letter-spacing: map-deep-get($typography, 'body-2', 'letter-spacing') !default;
$card-text-line-height: map-deep-get($typography, 'body-2', 'line-height') !default;
$card-text-padding-bottom: 16px !default;
$card-text-padding: 1rem !default;
$card-text-text-transform: map-deep-get($typography, 'body-2', 'text-transform') !default;
$card-tile-border-radius: 0 !default;
$card-title-comfortable-line-height: 1.75rem !default;
$card-title-compact-line-height: 1.55rem !default;
$card-title-font-size: map-deep-get($typography, 'h6', 'size') !default;
$card-title-font-weight: map-deep-get($typography, 'h6', 'weight') !default;
$card-title-hyphens: auto !default;
$card-title-letter-spacing: map-deep-get($typography, 'h6', 'letter-spacing') !default;
$card-title-line-height: map-deep-get($typography, 'h6', 'line-height') !default;
$card-title-overflow-wrap: anywhere !default;
$card-title-padding-top: 16px !default;
$card-title-padding: 1rem 1rem 0 !default;
$card-title-text-transform: none !default;
$card-title-word-break: normal !default;
$card-title-word-wrap: break-word !default;

$card-border: () !default;
$card-border: (
  $card-border-color,
  $card-border-style,
  $card-border-width
);

$card-title-densities: () !default;
$card-title-densities: (
  null: $card-title-line-height,
  'comfortable': $card-title-comfortable-line-height,
  'compact': $card-title-compact-line-height
);

$card-subtitle-densities: () !default;
$card-subtitle-densities: (
  null: $card-subtitle-line-height,
  'comfortable': $card-subtitle-comfortable-line-height,
  'compact': $card-subtitle-compact-line-height
);

$card-text-densities: () !default;
$card-text-densities: (
  null: $card-text-line-height,
  'comfortable': $card-text-comfortable-line-height,
  'compact': $card-text-compact-line-height
);

$card-theme: () !default;
$card-theme: (
  $card-background,
  $card-color
);
