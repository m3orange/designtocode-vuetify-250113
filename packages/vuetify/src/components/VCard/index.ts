export { default as VCard } from './VCard'
export { default as VCardActions } from './VCardActions'
export { default as VCardImg } from './VCardImg'
export { default as VCardSubtitle } from './VCardSubtitle'
export { default as VCardText } from './VCardText'
export { default as VCardTitle } from './VCardTitle'
