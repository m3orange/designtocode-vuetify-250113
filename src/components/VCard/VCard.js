require('../../stylus/components/_cards.styl')

import Colorable from '../../mixins/colorable'
import Routable from '../../mixins/routable'

export default {
  name: 'v-card',

  mixins: [Colorable, Routable],

  props: {
    flat: Boolean,
    height: {
      type: String,
      default: 'auto'
    },
    hover: Boolean,
    img: String,
    raised: Boolean,
    tag: {
      type: String,
      default: 'div'
    },
    tile: Boolean
  },

  computed: {
    classes () {
      return {
        'card': true,
        'card--flat': this.flat,
        'card--horizontal': this.horizontal,
        'card--hover': this.hover,
        'card--raised': this.raised,
        'card--tile': this.tile,
        'theme--light': this.light,
        'theme--dark': this.dark
      }
    },
    styles () {
      const style = {
        height: isNaN(this.height) ? this.height : `${this.height}px`
      }

      if (this.img) {
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h) {
    const { tag, data } = this.generateRouteLink()

    data.class = this._computedClasses
    data.style = this.styles

    return h(tag, data, this.$slots.default)
  }
}
