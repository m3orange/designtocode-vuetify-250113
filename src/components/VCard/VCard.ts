// Styles
import '../../stylus/components/_cards.styl'

// Mixins
import Colorable from '../../mixins/colorable'
import Elevatable from '../../mixins/elevatable'
import Routable from '../../mixins/routable'
import Themeable from '../../mixins/themeable'

// Helpers
import { convertToUnit } from '../../util/helpers'
import { deprecate } from '../../util/console'
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'
import { ClassesObject } from './../../../types'

/* @vue/component */
export default mixins(
  Colorable,
  Elevatable,
  Routable,
  Themeable
).extend({
  name: 'v-card',

  props: {
    elevation: {
      type: [Number, String],
      default: 1
    },
    flat: Boolean,
    height: [Number, String],
    /* @deprecated */
    img: String,
    maxHeight: [Number, String],
    maxWidth: [Number, String],
    /* @deprecated */
    raised: Boolean,
    tag: {
      type: String,
      default: 'div'
    },
    tile: Boolean,
    width: [Number, String]
  },

  computed: {
    classes (): ClassesObject {
      return this.addBackgroundColorClassChecks({
        'v-card': true,
        'v-card--flat': this.flat,
        'v-card--hover': Boolean(this.hover),
        'v-card--raised': this.raised,
        'v-card--tile': this.tile,
        ...this.themeClasses,
        ...this.elevationClasses
      })
    },
    computedElevation (): number | string {
      if (this.raised) {
        deprecate('<v-card raised>', '<v-card elevation="3">', this)

        return 3
      }

      if (this.hover === '' && this.isMouseOver) {
        deprecate('<v-card hover>', '<v-card hover="8">', this)

        return 8
      }

      return this.flat
        ? 0
        : Elevatable.options.computed.computedElevation.call(this)
    },
    styles (): object {
      const style: Record<string, any> = {
        height: convertToUnit(this.height),
        maxHeight: convertToUnit(this.maxHeight),
        maxWidth: convertToUnit(this.maxWidth),
        width: convertToUnit(this.width)
      }

      if (this.img) {
        deprecate('<v-card img="...">', 'a nested <v-img>', this)
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h): VNode {
    const { tag, data } = this.generateRouteLink()

    data.on = {
      ...data.on,
      ...this.listeners()
    }
    data.style = this.styles

    return h(tag, data, this.$slots.default)
  }
})
