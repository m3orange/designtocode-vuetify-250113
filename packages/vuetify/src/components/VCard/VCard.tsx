// Styles
import './VCard.sass'

// Components
import { VCardActions, VCardImg, VCardSubtitle, VCardText, VCardTitle } from './'

// Composables
import { makeBorderProps, useBorder } from '@/composables/border'
import { makeDimensionProps, useDimension } from '@/composables/dimensions'
import { makeElevationProps, useElevation } from '@/composables/elevation'
import { makePositionProps, usePosition } from '@/composables/position'
import { makeRoundedProps, useRounded } from '@/composables/rounded'
import { makeTagProps } from '@/composables/tag'
import { useBackgroundColor } from '@/composables/color'
import { useTheme } from '@/composables/theme'

// Utilities
import { defineComponent, toRef } from 'vue'
import { makeProps } from '@/util'

export default defineComponent({
  name: 'VCard',

  props: makeProps({
    color: String,
    flat: Boolean,
    hover: Boolean,
    image: String,
    raised: Boolean,
    subtitle: String,
    text: String,
    title: String,
    ...makeBorderProps(),
    ...makeDimensionProps(),
    ...makeElevationProps(),
    ...makePositionProps(),
    ...makeRoundedProps(),
    ...makeTagProps(),
  }),

  setup (props, { slots }) {
    const { themeClasses } = useTheme()
    const { backgroundColorClasses, backgroundColorStyles } = useBackgroundColor(toRef(props, 'color'))
    const { borderClasses } = useBorder(props, 'v-card')
    const { dimensionStyles } = useDimension(props)
    const { elevationClasses } = useElevation(props)
    const { positionClasses, positionStyles } = usePosition(props, 'v-card')
    const { roundedClasses } = useRounded(props, 'v-card')

    return () => {
      const hasImage = (slots.image || props.image)

      return (
        <props.tag
          class={[
            'v-card',
            {
              'v-card--flat': props.flat,
              'v-card--hover': props.hover,
              // 'v-card--link': props.isClickable,
              // 'v-card--loading': props.loading,
              // 'v-card--disabled': props.disabled,
              'v-card--raised': props.raised,
            },
            themeClasses.value,
            backgroundColorClasses.value,
            borderClasses.value,
            elevationClasses.value,
            positionClasses.value,
            roundedClasses.value,
          ]}
          style={[
            backgroundColorStyles.value,
            dimensionStyles.value,
            positionStyles.value,
          ]}
        >
          { hasImage && (
            <VCardImg>
              { slots.image
                ? slots.image?.({ src: props.image })
                : (<img src={ props.image } />)
              }
            </VCardImg>
          ) }

          { props.title && (
            <VCardTitle>{props.title}</VCardTitle>
          ) }

          { props.subtitle && (
            <VCardSubtitle>{props.subtitle}</VCardSubtitle>
          ) }

          { props.text && (
            <VCardText>{props.text}</VCardText>
          ) }

          { slots.default?.() }

          { slots.actions && (
            <VCardActions>{slots.actions()}</VCardActions>
          ) }
        </props.tag>
      )
    }
  },
})
