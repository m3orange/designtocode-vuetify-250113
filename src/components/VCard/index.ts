import { createSimpleFunctional } from '../../util/helpers'
import VCard from './VCard'
import VCardMedia from './VCardMedia'
import VCardTitle from './VCardTitle'
import Vue from 'vue'

const VCardActions = Vue.extend(createSimpleFunctional('v-card__actions'))
const VCardText = Vue.extend(createSimpleFunctional('v-card__text'))

export { VCard, VCardMedia, VCardTitle, VCardActions, VCardText }

VCard.$_vuetify_subcomponents = {
  VCardMedia,
  VCardTitle,
  VCardActions,
  VCardText
}

export default VCard
