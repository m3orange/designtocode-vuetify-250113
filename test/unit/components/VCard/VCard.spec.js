import { test } from '@/test'
import VCard from '@/components/VCard'

const imgDeprecate = `'<v-card img="...">' is deprecated, use 'a nested <v-img>' instead`
const raisedDeprecate = `'<v-card raised>' is deprecated, use '<v-card elevation="3">' instead`

test('VCard.vue', ({ mount }) => {
  it('should render component and match snapshot', () => {
    const wrapper = mount(VCard)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render card with img', () => {
    const wrapper = mount(VCard, {
      propsData: {
        img: 'image.jpg'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(imgDeprecate).toHaveBeenTipped()
  })

  it('should render a flat card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        flat: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a raised card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        raised: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(raisedDeprecate).toHaveBeenTipped()
  })

  it('should render a colored card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        color: 'blue lighten-1'
      }
    })

    expect(wrapper.element.classList).toContain('blue')
    expect(wrapper.element.classList).toContain('lighten-1')
  })

  it('should render a tile card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        tile: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a card with custom height', () => {
    const heightpx = '400px'
    const wrapper = mount(VCard, {
      propsData: {
        height: heightpx
      }
    })

    expect(wrapper.hasStyle('height', heightpx)).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      height: 401
    })
    expect(wrapper.hasStyle('height', '401px')).toBe(true)
  })

  it('should render a tile card', () => {
    const wrapper = mount(VCard, {
      propsData: {
        tile: true
      }
    })

    expect(wrapper.hasClass('v-card--tile')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
