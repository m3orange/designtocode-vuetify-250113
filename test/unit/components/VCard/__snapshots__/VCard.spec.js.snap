// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCard.vue should render a card with custom height 1`] = `

<div class="v-card elevation-1"
     style="height: 400px;"
>
</div>

`;

exports[`VCard.vue should render a flat card 1`] = `

<div class="v-card v-card--flat elevation-0">
</div>

`;

exports[`VCard.vue should render a raised card 1`] = `

<div class="v-card v-card--raised elevation-3">
</div>

`;

exports[`VCard.vue should render a tile card 1`] = `

<div class="v-card v-card--tile elevation-1">
</div>

`;

exports[`VCard.vue should render a tile card 2`] = `

<div class="v-card v-card--tile elevation-1">
</div>

`;

exports[`VCard.vue should render card with img 1`] = `

<div class="v-card elevation-1">
</div>

`;

exports[`VCard.vue should render component and match snapshot 1`] = `

<div class="v-card elevation-1">
</div>

`;
